_posts/2018-02-05-unrealiable-network.md

```diff
@@ -0,0 +1,111 @@
+---
+layout: post
+title: Quick Tips for App Developers on Surviving with Unrealiable Network
+author: jarzka
+excerpt: Methods that you, application developer, can do to make your app feel more stable even if the underlying network is unreliable.
+tags:
+- clojure
+- clojurescript
+- network programming
+---
+
+You cannot always choose where your application will be hosted. Sometimes the network infrastructure can be unrealiable, meaning that network requests can sometimes take a long time to complete and requests will fail at random rates. Even if these problems could not be prevented from happening, luckily we, application developers, have some tricks we can use to make our applications *feel* more stable in unrealiable network. And even if your network infrastucture is solid, it is not guaranteed that occasional communication failures won't happen, as illustrated in [Two Generals Problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem). Thus, these quick tips, targeted mainly for single page web application and mobile application developers, will probably help your application behave more stable and quickly for your end users.
+
+# Avoid Making Network Requests as Much as You Can
+
+The best way to avoid getting failing network requests is of course to abandom making them completely.
+
+![Failing network requests meme](/img/unrealiable-network/network_requests_meme.jpg)
+
+Thought this is very efficient, it is not very practical for most of us. Majority of web and mobile apps depend on getting data from servers, and thus, making network requests is mandatory. Still, if we cannot abandom them completely, we can at least reduce the amount of them.
+
+How can we reduce the amount of network requests? First, we should analyse what kind of requests our application is making. The developer tools of [Chrome](https://developer.chrome.com/devtools) and [Firefox](https://developer.mozilla.org/fi/docs/Tools) are very practical for this analysis. On the network tab, you can see all the requests your application is making.
+
+![Chrome Network Tab](/img/unrealiable-network/chrome_network_tab.png)
+
+The way to reduce the amount of requests depends on your application. The following questions can help you to find the requests that could be improved:
+
+**Are you making two separate network requests for saving data and then getting the same data again from the server to update the view?** 
```

Comment:
Are you saying that CQRS is bad? I wouldn't claim this as a general rule. Depending on your app architecture and the type of failures, it may be better to have a smaller request to save something go through and then failing the fetch than to combine them.

For example saving something and making an optimistic update of the app state before we get the fetch response.

Perhaps expand on this point.

Reply:
Good suggestion. Will be added.